snuba/clickhouse/formatter/query.py

```diff
@@ -48,39 +51,37 @@ def format_query(query: FormattableQuery, settings: RequestSettings) -> Formatte
     return FormattedQuery(_format_query_content(query))
 
 
-def format_data_source(data_source: DataSource) -> FormattedNode:
+class DataSourceFormatter(DataSourceVisitor[FormattedNode, Table]):
     """
     Builds a FormattedNode out of any type of DataSource object. This
     is called when formatting the FROM clause of every nested query
     in a Composite query.
-
-    Remark: The ideal way to structure this would be a visitor just
-    like we did for the AST expression but this is not possible
-    without introducing either circular dependencies or dropping
-    type checking since the visitor and the classes to visit cannot
-    be in the same module as of now and in a visitor pattern they would
-    have to be mutually dependent.
     """
-    if isinstance(data_source, Table):
-        return format_table(data_source)
-    elif isinstance(data_source, Query):
-        return FormattedSubQuery(_format_query_content(data_source))
-    elif isinstance(data_source, JoinClause):
-        return data_source.accept(JoinFormatter())
-    elif isinstance(data_source, CompositeQuery):
-        return FormattedSubQuery(_format_query_content(data_source))
-    else:
-        raise NotImplementedError(f"Unsupported data source {type(data_source)}")
 
+    def _visit_simple_source(self, data_source: Table) -> StringNode:
+        """
+        Formats a simple table data source.
+        """
 
-def format_table(data_source: Table) -> StringNode:
-    """
-    Formats a simple table data source.
-    """
+        final = " FINAL" if data_source.final else ""
+        sample = (
+            f" SAMPLE {data_source.sampling_rate}" if data_source.sampling_rate else ""
+        )
```

Comment:
We may have discussed this but remind me again - are final and sample just getting applied to each one of the referenced data sources independently?

Reply:
Yes they are. As per clickhouse docs FINAL and SAMPLE are property of the subquery. https://clickhouse.tech/docs/en/sql-reference/statements/select/from/#select-from

Reply:
I mean for us sample and final are in the main query, so are we just going to push it down to every subquery equally?

Reply:
They are associated with each entity in SnQL they are not present in the main query. The old language does not support composite queries.